### DISCUSSION TITLE
Default values that are visible in projenrc

### DISCUSSION THREAD

**Discussion by automartin5000** [projen/projen#3195] [Q&A]
How do I make some of the options in my project actual code. For example, I'm extending the `AwsCdkTypeScriptApp` project type and I noticed that options such as `projenrcTs` and `defaultReleaseBranch` are automatically set but also populated as actual params. While other options are commented out (I've discovered from the `@featured` flag). I've added a couple of my own options as well as included some dependencies when sample code is generated. In those cases, I'd like the extra dependencies to be explicitly defined as `deps` and the options to be uncommented.

Hope that makes sense, let me know if you need code examples.

**Comment by mrgrain** [comment]
You are looking for `@pjnew`

https://github.com/projen/projen/blob/main/src%2Ftypescript%2Ftypescript.ts#L138

**Reply by automartin5000** [reply]
Cool thanks! And for my deps, if I want the ones I've added via `addDeps` to also show up. Does that mean I need to create my own prop for `deps` and set `pjnew` on it? I'm guessing yes since `addDeps` only accepts an array of strings with no other options.

**Reply by mrgrain** [reply]
> Cool thanks! And for my deps, if I want the ones I've added via addDeps to also show up. Does that mean I need to create my own prop for deps and set pjnew on it? I'm guessing yes since addDeps only accepts an array of strings with no other options.

You currently can't make method calls show up in the generated projenrc file, only options. So yes, you'd have to include a `deps` option in your project type.

Saying that, anything that is `@pjnew` is just a suggested default that can be changed by the user. `addDeps` is the way to go to make sure a dependency is definitely added and don't leave the user a  choice (ignoring escape hatches).

**Reply by automartin5000** [reply]
I'm trying to show an example of a dependency to go along with my sample code. But if the user removes the dependency in the runtime code, I want them to be able to remove the dependency from projen.

Also, I just tried `@pjnew` and now creating a new project is yelling at me with:
`required option "<option>" with a @default must use primitive types`. I'm using a type from an `enum`. Is that not allowed?

**Reply by automartin5000** [reply]
And just FYI, now I'm getting this during build:
```
Interface "CustomAwsCdkTypescriptAppProps" re-declares member "deps". This is not supported as it results in invalid C#
```
I think I might be blocked here and have to drop the dependency from sample code 🤔

**Reply by mrgrain** [reply]
> And just FYI, now I'm getting this during build:
> ```
> Interface "CustomAwsCdkTypescriptAppProps" re-declares member "deps". This is not supported as it results in invalid C#
> ```
> I think I might be blocked here and have to drop the dependency from sample code 🤔 

This I know. That's a jsii limitation. Which obviously makes your life harder when trying to do things like `@pjnew`.

The usual way to get around these jsii restrictions is to create an entirely new interface. Some people use [jsii-struct-builder](https://github.com/mrgrain/jsii-struct-builder) to achieve this more easily.

**Reply by mrgrain** [reply]
> Also, I just tried @pjnew and now creating a new project is yelling at me with:
required option "<option>" with a @default must use primitive types. I'm using a type from an enum. Is that not allowed?

Sounds like it? I'd have to look it up in the code. Enums are supported as cli arguments, so I don't see a reason why they can't using `@default`

**Reply by automartin5000** [reply]
@mrgrain Looks like `@pjnew` doesn't allow enums: https://github.com/projen/projen/blob/6bd3e0d99f0a10a933f53630d7eeb98e378047b9/src/inventory.ts#L466

So I reverted to `@featured` which isn't as nice as `@pjnew`, but will probably work well enough for now.

**Reply by mrgrain** [reply]
> @mrgrain Looks like `@pjnew` doesn't allow enums: https://github.com/projen/projen/blob/6bd3e0d99f0a10a933f53630d7eeb98e378047b9/src/inventory.ts#L466

Well that's not great. We should fix that. 😅

**Comment by automartin5000** [comment]
@mrgrain Just following up on this as it relates to `deps`. When I use `addDeps` or `addDevDeps`, dependencies are added invisibly to the user and it doesn't appear like there's any way for the user to override them. Adding a different version of the same dependency via `deps` or `devDeps` in the end user's project appears to have no effect. Does this sound accurate? If so, I'd propose adding a `pjnew` option in those methods.

**Reply by mrgrain** [reply]
You can only have `@pjnew` on options. I suppose it would be possible to support method calls, but I'm not sure if it's worth the effort.

If you don't want these dependencies to be forced, why are you using `addDeps`?

**Reply by automartin5000** [reply]
@mrgrain It's an internal CDK library, it's used with sample code. I did just discover that on the generated project I can call `project.addDeps` and that overrides the default version, but I still think there should be a way to optionally make pre-installed deps clearer to the end-user and easier to maintain.
